numpy/core/src/multiarray/datetime.c

```diff
@@ -430,6 +431,105 @@ npy_int64 extract_unit(npy_datetime *d, npy_datetime unit) {
     return div;
 }
 
+/*
+ * Convert datetime64 to a tuple (quad_centuries, microseconds, attoseconds)
+ * then hash the tuple.
+ * quad_centuries, because it's easy to convert both (number_of_years) and
+ * (number_of_days) to (number_of_400_year_periods).
+ * microseconds, because 400 years of microseconds fits into 54 bits.
+ * attoseconds, because that's the highest resolution.
+ */
+NPY_NO_EXPORT npy_int64
+datetime_hash(PyArray_DatetimeMetaData *meta, npy_datetime dt)
+{
+    npy_datetimestruct dts;
+    npy_int64 quad_centuries;
+    npy_int64 microseconds;
+    npy_int64 attoseconds;
+
+    dts.year = 1970;
+    dts.month = 1;
+    dts.day = 1;
+
+    /* TODO: Change to a mechanism that avoids the potential overflow */
+    dt *= meta->num;
+
+    switch (meta->base) {
+    case NPY_FR_Y:
+        quad_centuries = extract_unit(&dt, 400);
+        dts.year = 1970 + dt;
+        microseconds = get_datetimestruct_days(&dts) * 24LL * 60 * 60 * 1000 * 1000;
+        attoseconds = 0LL;
+        break;
+    case NPY_FR_M:
+        quad_centuries = extract_unit(&dt, 12 * 400);
+        dts.year = 1970 + extract_unit(&dt, 12);
+        dts.month = dt + 1;
+        microseconds = get_datetimestruct_days(&dts) * 24LL * 60 * 60 * 1000 * 1000;
+        attoseconds = 0LL;
+        break;
+    case NPY_FR_W:
+        quad_centuries = extract_unit(&dt, DAYS_IN_400_YEARS / 7);
+        microseconds = dt * 7 * 24LL * 60 * 60 * 1000 * 1000;
```

Comment:
This looks questionable to me - is this division exact?

Reply:
You mean float(DAYS_IN_400_YEARS) / 7.0 == 20871.0 ? Yep, it's exact.